In [1]:
# Import libraries

import pandas as pd
import numpy as np
import os

## Importing Data

In [2]:
#import dataframes
df = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\original data\orders.csv', index_col=False)
path = r'C:\Users\amany\A4-Instacart Basket Analysis'
df_prods = pd.read_csv(os.path.join(path, '02 Data' , 'prepared data' , 'products_cleaned.csv'))
df_ords = pd.read_csv(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\original data\orders.csv')
ords_prods_merged_updated = pd.read_pickle(r'C:\Users\amany\A4-Instacart Basket Analysis\02 Data\prepared data\ords_prods_merged_updated.pkl')

## part 1 : working on a subset

In [3]:
# Create a subset of first million rows

df =ords_prods_merged_updated[:1000000]

In [4]:
df.shape

(1000000, 20)

In [5]:
df.head(10)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0   2539329        1    prior             1                   2   
1   2398795        1    prior             2                   3   
2    473747        1    prior             3                   3   
3   2254736        1    prior             4                   4   
4    431534        1    prior             5                   4   
5   3367565        1    prior             6                   2   
6    550135        1    prior             7                   1   
7   3108588        1    prior             8                   1   
8   2295261        1    prior             9                   1   
9   2550362        1    prior            10                   4   

   order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                  8                     NaN         196                  1   
1                  7                    15.0         196                  1   
2                 12                    21.0         196                  1   
3                  7                    29.0         196                  1   
4                 15                    28.0         196                  1   
5                  7                    19.0         196                  1   
6                  9                    20.0         196                  1   
7                 14                    14.0         196                  2   
8                 16                     0.0         196                  4   
9                  8                    30.0         196                  1   

   reordered _merge product_name  aisle_id  department_id  prices match  \
0          0   both         Soda        77              7     9.0  both   
1          1   both         Soda        77              7     9.0  both   
2          1   both         Soda        77              7     9.0  both   
3          1   both         Soda        77              7     9.0  both   
4          1   both         Soda        77              7     9.0  both   
5          1   both         Soda        77              7     9.0  both   
6          1   both         Soda        77              7     9.0  both   
7          1   both         Soda        77              7     9.0  both   
8          1   both         Soda        77              7     9.0  both   
9          1   both         Soda        77              7     9.0  both   

     price_range_loc     busiest_day    busiest_days busiest_period_of_day  
0  Mid-range product  Regularly busy  Regularly busy        Average orders  
1  Mid-range product  Regularly busy    Slowest days        Average orders  
2  Mid-range product  Regularly busy    Slowest days           Most orders  
3  Mid-range product      Least busy    Slowest days        Average orders  
4  Mid-range product      Least busy    Slowest days           Most orders  
5  Mid-range product  Regularly busy  Regularly busy        Average orders  
6  Mid-range product  Regularly busy    Busiest days           Most orders  
7  Mid-range product  Regularly busy    Busiest days           Most orders  
8  Mid-range product  Regularly busy    Busiest days           Most orders  
9  Mid-range product      Least busy    Slowest days        Average orders

In [6]:
#group a given dataframe by a given column
df.groupby('product_name')

In [7]:
#Split the data into groups based on “department_id.”
#Apply the agg() function to each group to obtain the mean values for the “order_number” column.

df.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
4                18.825780
7                17.472355
13               17.993423
14               19.246334
16               19.463012
17               11.294069
19               19.305237
20               17.599636

In [8]:
# conduct the aggregations without use of the agg() function.
df.groupby('department_id')['order_number'].mean()


department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [9]:
df.groupby('department_id').order_number.mean()


department_id
4     18.825780
7     17.472355
13    17.993423
14    19.246334
16    19.463012
17    11.294069
19    19.305237
20    17.599636
Name: order_number, dtype: float64

In [10]:
#producing not only the mean but also the min and max on the order number column
df.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
4                18.825780   1  99
7                17.472355   1  99
13               17.993423   1  99
14               19.246334   1  99
16               19.463012   1  99
17               11.294069   1  98
19               19.305237   1  99
20               17.599636   1  99

#### creating a flag for customers with many orders. to fing the loyal customers

Split the data into groups based on the “user_id” column.
Apply the transform() function on the “order_number” column to generate the maximum orders for each user.
Create a new column, “max_order,” into which you’ll place the results of your aggregation.


In [11]:
df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)

C:\Users\amany\AppData\Local\Temp\ipykernel_1988\3693603773.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['max_order'] = df.groupby(['user_id'])['order_number'].transform(np.max)


In [12]:
df.head(15)

order_id  user_id eval_set  order_number  orders_day_of_week  \
0    2539329        1    prior             1                   2   
1    2398795        1    prior             2                   3   
2     473747        1    prior             3                   3   
3    2254736        1    prior             4                   4   
4     431534        1    prior             5                   4   
5    3367565        1    prior             6                   2   
6     550135        1    prior             7                   1   
7    3108588        1    prior             8                   1   
8    2295261        1    prior             9                   1   
9    2550362        1    prior            10                   4   
10   2968173       15    prior            15                   1   
11   1870022       15    prior            17                   2   
12   1911383       15    prior            18                   2   
13   2715276       15    prior            21                   1   
14    487368       15    prior            22                   1   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   7                    15.0         196                  1   
2                  12                    21.0         196                  1   
3                   7                    29.0         196                  1   
4                  15                    28.0         196                  1   
5                   7                    19.0         196                  1   
6                   9                    20.0         196                  1   
7                  14                    14.0         196                  2   
8                  16                     0.0         196                  4   
9                   8                    30.0         196                  1   
10                  9                     7.0         196                  2   
11                 16                     8.0         196                  6   
12                 11                     7.0         196                  1   
13                  9                     7.0         196                  1   
14                 10                    14.0         196                  1   

    reordered  ... product_name aisle_id  department_id  prices  match  \
0           0  ...         Soda       77              7     9.0   both   
1           1  ...         Soda       77              7     9.0   both   
2           1  ...         Soda       77              7     9.0   both   
3           1  ...         Soda       77              7     9.0   both   
4           1  ...         Soda       77              7     9.0   both   
5           1  ...         Soda       77              7     9.0   both   
6           1  ...         Soda       77              7     9.0   both   
7           1  ...         Soda       77              7     9.0   both   
8           1  ...         Soda       77              7     9.0   both   
9           1  ...         Soda       77              7     9.0   both   
10          0  ...         Soda       77              7     9.0   both   
11          1  ...         Soda       77              7     9.0   both   
12          1  ...         Soda       77              7     9.0   both   
13          1  ...         Soda       77              7     9.0   both   
14          1  ...         Soda       77              7     9.0   both   

      price_range_loc     busiest_day    busiest_days busiest_period_of_day  \
0   Mid-range product  Regularly busy  Regularly busy        Average orders   
1   Mid-range product  Regularly busy    Slowest days        Average orders   
2   Mid-range product  Regularly busy    Slowest days           Most orders   
3   Mid-range product      Least busy    Slowest days        Average orders   
4   Mid-range product      Least busy    Slowest days 

#### step 3 Deriving Columns with loc()


In [13]:
# create a flag that assigns a “loyalty” label to a user ID based on its corresponding max order value
df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


C:\Users\amany\AppData\Local\Temp\ipykernel_1988\1961839803.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


In [14]:
df.loc[(df['max_order'] <= 40) & (df['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'


In [15]:
df.loc[df['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [16]:
df['loyalty_flag'].value_counts(dropna = False)

Regular customer    463844
Loyal customer      335086
New customer        201070
Name: loyalty_flag, dtype: int64

In [17]:
df.shape

(1000000, 22)

In [18]:
df

order_id  user_id eval_set  order_number  orders_day_of_week  \
0        2539329        1    prior             1                   2   
1        2398795        1    prior             2                   3   
2         473747        1    prior             3                   3   
3        2254736        1    prior             4                   4   
4         431534        1    prior             5                   4   
...          ...      ...      ...           ...                 ...   
999995   2836489   164627    prior            14                   0   
999996   1843600   164632    prior             5                   1   
999997    733106   164632    prior             9                   2   
999998   1650124   164632    prior            17                   6   
999999   3378258   164632    prior            22                   6   

        order_hour_of_day  days_since_prior_order  product_id  \
0                       8                     NaN         196   
1                       7                    15.0         196   
2                      12                    21.0         196   
3                       7                    29.0         196   
4                      15                    28.0         196   
...                   ...                     ...         ...   
999995                 15                     6.0       30489   
999996                 19                     9.0       30489   
999997                 22                    10.0       30489   
999998                 17                    13.0       30489   
999999                 10                    15.0       30489   

        add_to_cart_order  reordered  ... aisle_id department_id  prices  \
0                       1          0  ...       77             7     9.0   
1                       1          1  ...       77             7     9.0   
2                       1          1  ...       77             7     9.0   
3                       1          1  ...       77             7     9.0   
4                       1          1  ...       77             7     9.0   
...                   ...        ...  ...      ...           ...     ...   
999995                  2          0  ...       67            20     7.5   
999996                  2          0  ...       67            20     7.5   
999997                  3          1  ...       67            20     7.5   
999998                  1          1  ...       67            20     7.5   
999999                  3          1  ...       67            20     7.5   

        match    price_range_loc     busiest_day    busiest_days  \
0        both  Mid-range product  Regularly busy  Regularly busy   
1        both  Mid-range product  Regularly busy    Slowest days   
2        both  Mid-range product  Regularly busy    Slowest days   
3        both  Mid-range product      Least busy    Slowest days   
4        both  Mid-range product      Least busy    Slowest days   
...       ...                ...             ...             ...   
999995   both  Mid-range product     Busiest day    Busiest days   
999996   both  Mid-range product  Regularly busy    Busiest days   
999997   both  Mid-range product  Regularly busy  Regularly busy   
999998   both  Mid-range product  Regularly busy  Regularly busy   
999999   both  Mid-range product  Regularly busy  Regularly busy   

       busiest_period_of_day max_order      loyalty_flag  
0             Average orders        10      New customer  
1             Average orders        10      New customer  
2                Most orders        10      New customer  
3             Average orders        10      New customer  
4                Most orders        10      New customer  
...                      ...       ...               ...  
999995           Most orders        23  Regular customer  
999996        Average orders        31  Regular customer  
999997        Average orders        31  Regular customer  
999998        Average orders        31  Re

## part 2, working on the entire dataframe

In [19]:
ords_prods_merged_updated

order_id  user_id eval_set  order_number  orders_day_of_week  \
0          2539329        1    prior             1                   2   
1          2398795        1    prior             2                   3   
2           473747        1    prior             3                   3   
3          2254736        1    prior             4                   4   
4           431534        1    prior             5                   4   
...            ...      ...      ...           ...                 ...   
32434207   1320836   202557    prior            17                   2   
32434208     31526   202557    prior            18                   5   
32434209    758936   203436    prior             1                   2   
32434210   2745165   203436    prior             2                   3   
32434211   3093936   205420    prior             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32434207                 15                     1.0       43553   
32434208                 11                     3.0       43553   
32434209                  7                     NaN       42338   
32434210                  5                    15.0       42338   
32434211                 14                     NaN       28818   

          add_to_cart_order  reordered _merge                   product_name  \
0                         1          0   both                           Soda   
1                         1          1   both                           Soda   
2                         1          1   both                           Soda   
3                         1          1   both                           Soda   
4                         1          1   both                           Soda   
...                     ...        ...    ...                            ...   
32434207                  2          1   both            Orange Energy Shots   
32434208                  2          1   both            Orange Energy Shots   
32434209                  4          0   both          Zucchini Chips, Pesto   
32434210                 16          1   both          Zucchini Chips, Pesto   
32434211                  8          0   both  Hot Oatmeal Multigrain Raisin   

          aisle_id  department_id  prices match    price_range_loc  \
0               77              7     9.0  both  Mid-range product   
1               77              7     9.0  both  Mid-range product   
2               77              7     9.0  both  Mid-range product   
3               77              7     9.0  both  Mid-range product   
4               77              7     9.0  both  Mid-range product   
...            ...            ...     ...   ...                ...   
32434207        64              7     3.7  both  Low-range product   
32434208        64              7     3.7  both  Low-range product   
32434209        50             19     6.9  both  Mid-range product   
32434210        50             19     6.9  both  Mid-range product   
32434211       130             14    10.3  both  Mid-range product   

             busiest_day    busiest_days busiest_period_of_day  
0         Regularly busy  Regularly busy        Average orders  
1         Regularly busy    Slowest days        Average orders  
2         Regularly busy    Slowest days           Most orders  
3             Least busy    Slowest days        Average orders  
4             Least busy    Slowest days           Most orders  
...                  ...             ...                   ...  
32434207  Regularly busy  Regularly busy           Most ord

In [20]:
#group a given dataframe by a given column
ords_prods_merged_updated.groupby('product_name')

In [21]:
# Grouping by 'department_id' and 'order_number', and Calculating Mean using agg() Function
ords_prods_merged_updated.groupby('department_id').agg({'order_number' : ['mean']})

order_number
                      mean
department_id             
1                15.457687
2                17.277920
3                17.179756
4                17.811403
5                15.213779
6                16.439806
7                17.225773
8                15.340520
9                15.895474
10               20.197148
11               16.170828
12               15.887622
13               16.583304
14               16.757377
15               16.165037
16               17.663250
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

#### insights, In the entire dataframe, the order_number mean across all department_ids is generally distributed between approximately 15 and 20, with some outliers like department 21 having a significantly higher mean (22.90) and department 17 slightly lower (15.69). In the subset dataframe, the order_number mean is skewed slightly higher for most department_ids included in the subset, with some exceeding 19 (e.g., departments 4, 14, 16, and 19).

### conduct the aggregations without use of the agg() function.
ords_prods_merged_updated.groupby('department_id')['order_number'].mean()

In [22]:
ords_prods_merged_updated.groupby('department_id').order_number.mean()


department_id
1     15.457687
2     17.277920
3     17.179756
4     17.811403
5     15.213779
6     16.439806
7     17.225773
8     15.340520
9     15.895474
10    20.197148
11    16.170828
12    15.887622
13    16.583304
14    16.757377
15    16.165037
16    17.663250
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

In [23]:
#producing not only the mean but also the min and max on the order number column
ords_prods_merged_updated.groupby('department_id').agg({'order_number': ['mean', 'min', 'max']})

order_number        
                      mean min max
department_id                     
1                15.457687   1  99
2                17.277920   1  99
3                17.179756   1  99
4                17.811403   1  99
5                15.213779   1  99
6                16.439806   1  99
7                17.225773   1  99
8                15.340520   1  99
9                15.895474   1  99
10               20.197148   1  99
11               16.170828   1  99
12               15.887622   1  99
13               16.583304   1  99
14               16.757377   1  99
15               16.165037   1  99
16               17.663250   1  99
17               15.694469   1  99
18               19.310397   1  99
19               17.177343   1  99
20               16.473447   1  99
21               22.902379   1  99

In [24]:
# creating a flag for customers with many orders. to fing the loyal customers
ords_prods_merged_updated['max_order'] = ords_prods_merged_updated.groupby(['user_id'])['order_number'].transform(np.max)

In [25]:
ords_prods_merged_updated.loc[ords_prods_merged_updated['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'


In [26]:
ords_prods_merged_updated.loc[(ords_prods_merged_updated['max_order'] <= 40) & (ords_prods_merged_updated['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'


In [27]:
ords_prods_merged_updated.loc[ords_prods_merged_updated['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [28]:
ords_prods_merged_updated['loyalty_flag'].value_counts(dropna = False)

Regular customer    15891077
Loyal customer      10293737
New customer         6249398
Name: loyalty_flag, dtype: int64

In [29]:
ords_prods_merged_updated.shape

(32434212, 22)

In [30]:
ords_prods_merged_updated

order_id  user_id eval_set  order_number  orders_day_of_week  \
0          2539329        1    prior             1                   2   
1          2398795        1    prior             2                   3   
2           473747        1    prior             3                   3   
3          2254736        1    prior             4                   4   
4           431534        1    prior             5                   4   
...            ...      ...      ...           ...                 ...   
32434207   1320836   202557    prior            17                   2   
32434208     31526   202557    prior            18                   5   
32434209    758936   203436    prior             1                   2   
32434210   2745165   203436    prior             2                   3   
32434211   3093936   205420    prior             1                   4   

          order_hour_of_day  days_since_prior_order  product_id  \
0                         8                     NaN         196   
1                         7                    15.0         196   
2                        12                    21.0         196   
3                         7                    29.0         196   
4                        15                    28.0         196   
...                     ...                     ...         ...   
32434207                 15                     1.0       43553   
32434208                 11                     3.0       43553   
32434209                  7                     NaN       42338   
32434210                  5                    15.0       42338   
32434211                 14                     NaN       28818   

          add_to_cart_order  reordered  ... aisle_id department_id  prices  \
0                         1          0  ...       77             7     9.0   
1                         1          1  ...       77             7     9.0   
2                         1          1  ...       77             7     9.0   
3                         1          1  ...       77             7     9.0   
4                         1          1  ...       77             7     9.0   
...                     ...        ...  ...      ...           ...     ...   
32434207                  2          1  ...       64             7     3.7   
32434208                  2          1  ...       64             7     3.7   
32434209                  4          0  ...       50            19     6.9   
32434210                 16          1  ...       50            19     6.9   
32434211                  8          0  ...      130            14    10.3   

          match    price_range_loc     busiest_day    busiest_days  \
0          both  Mid-range product  Regularly busy  Regularly busy   
1          both  Mid-range product  Regularly busy    Slowest days   
2          both  Mid-range product  Regularly busy    Slowest days   
3          both  Mid-range product      Least busy    Slowest days   
4          both  Mid-range product      Least busy    Slowest days   
...         ...                ...             ...             ...   
32434207   both  Low-range product  Regularly busy  Regularly busy   
32434208   both  Low-range product  Regularly busy  Regularly busy   
32434209   both  Mid-range product  Regularly busy  Regularly busy   
32434210   both  Mid-range product  Regularly busy    Slowest days   
32434211   both  Mid-range product      Least busy    Slowest days   

         busiest_period_of_day max_order      loyalty_flag  
0               Average orders        10      New customer  
1               Average orders        10      New customer  
2                  Most orders        10      New customer  
3               Average orders        10      New customer  
4                  Most orders        10      New customer  
...                        ...       ...               ...  
32434207           Most orders        31  Regular customer  
32434208           Most orders        31  Regular 

### question 5

In [31]:
outliers = ords_prods_merged_updated[ords_prods_merged_updated['prices'] == 99999.0]


In [32]:
outlier_count = len(outliers)
total_entries = len(df)
outlier_percentage = (outlier_count / total_entries) * 100
print(f"Outliers represent {outlier_percentage:.2f}% of the data.")


Outliers represent 0.07% of the data.


In [33]:
ords_prods_merged_updated = ords_prods_merged_updated[ords_prods_merged_updated['prices'] != 99999.0]


In [34]:
print(ords_prods_merged_updated['prices'].describe())


count    3.243351e+07
mean     9.829328e+00
std      1.740661e+02
min      1.000000e+00
25%      4.200000e+00
50%      7.400000e+00
75%      1.130000e+01
max      1.490000e+04
Name: prices, dtype: float64


In [35]:
# Determining Basic Statistics of Spending Habits by Loyalty Category
ords_prods_merged_updated.groupby('loyalty_flag').agg({'prices' : ['mean', 'min', 'max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer     9.582517  1.0  14900.0
New customer      10.062416  1.0  14900.0
Regular customer   9.897543  1.0  14900.0

insights
Spending Habits by Loyalty Group:

Loyal Customers spend the least on average ($9.58), indicating a preference for cost-effective or discounted items.

New Customers spend the most on average ($10.06), possibly due to exploring premium products during initial interactions.

Regular Customers fall in between ($9.90), showing moderate spending behavior.

## question 6

In [36]:
# Creating column for average spender
ords_prods_merged_updated['average_spend']=ords_prods_merged_updated.groupby(['user_id'])['prices'].transform(np.mean)

C:\Users\amany\AppData\Local\Temp\ipykernel_1988\1598223287.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ords_prods_merged_updated['average_spend']=ords_prods_merged_updated.groupby(['user_id'])['prices'].transform(np.mean)


In [37]:
# Creating flag for low and high spender
ords_prods_merged_updated.loc[ords_prods_merged_updated['average_spend']<10,'spender_flag']='Low_spender'

C:\Users\amany\AppData\Local\Temp\ipykernel_1988\4069226224.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ords_prods_merged_updated.loc[ords_prods_merged_updated['average_spend']<10,'spender_flag']='Low_spender'


In [38]:
ords_prods_merged_updated.loc[ords_prods_merged_updated['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [39]:
# Check spender flags
ords_prods_merged_updated['spender_flag'].value_counts(dropna = False)

Low_spender     31855555
High_spender      577959
Name: spender_flag, dtype: int64

In [50]:
ords_prods_merged_updated[['user_id', 'average_spend', 'spender_flag', 'prices']].head(100)

user_id  average_spend  spender_flag  prices
0         1       6.367797   Low_spender     9.0
1         1       6.367797   Low_spender     9.0
2         1       6.367797   Low_spender     9.0
3         1       6.367797   Low_spender     9.0
4         1       6.367797   Low_spender     9.0
5         1       6.367797   Low_spender     9.0
6         1       6.367797   Low_spender     9.0
7         1       6.367797   Low_spender     9.0
8         1       6.367797   Low_spender     9.0
9         1       6.367797   Low_spender     9.0
10       15       3.980556   Low_spender     9.0
11       15       3.980556   Low_spender     9.0
12       15       3.980556   Low_spender     9.0
13       15       3.980556   Low_spender     9.0
14       15       3.980556   Low_spender     9.0
15       19       8.372059   Low_spender     9.0
16       19       8.372059   Low_spender     9.0
17       19       8.372059   Low_spender     9.0
18       21       7.012683   Low_spender     9.0
19       31       8.137793   Low_spender     9.0
20       31       8.137793   Low_spender     9.0
21       43       7.444898   Low_spender     9.0
22       43       7.444898   Low_spender     9.0
23       52       6.408876   Low_spender     9.0
24       52       6.408876   Low_spender     9.0
25       52       6.408876   Low_spender     9.0
26       52       6.408876   Low_spender     9.0
27       52       6.408876   Low_spender     9.0
28       52       6.408876   Low_spender     9.0
29       52       6.408876   Low_spender     9.0
30       52       6.408876   Low_spender     9.0
31       52       6.408876   Low_spender     9.0
32       52       6.408876   Low_spender     9.0
33       52       6.408876   Low_spender     9.0
34       52       6.408876   Low_spender     9.0
35       52       6.408876   Low_spender     9.0
36       52       6.408876   Low_spender     9.0
37       67       7.656790   Low_spender     9.0
38       67       7.656790   Low_spender     9.0
39       67       7.656790   Low_spender     9.0
40       67       7.656790   Low_spender     9.0
41       67       7.656790   Low_spender     9.0
42       67       7.656790   Low_spender     9.0
43       67       7.656790   Low_spender     9.0
44       67       7.656790   Low_spender     9.0
45       67       7.656790   Low_spender     9.0
46       67       7.656790   Low_spender     9.0
47       67       7.656790   Low_spender     9.0
48       67       7.656790   Low_spender     9.0
49       67       7.656790   Low_spender     9.0
50       67       7.656790   Low_spender     9.0
51       67       7.656790   Low_spender     9.0
52       67       7.656790   Low_spender     9.0
53       67       7.656790   Low_spender     9.0
54       67       7.656790   Low_spender     9.0
55       67       7.656790   Low_spender     9.0
56       81       8.262963   Low_spender     9.0
57       81       8.262963   Low_spender     9.0
58       82       6.441221   Low_spender     9.0
59       82       6.441221   Low_spender     9.0
60       82       6.441221   Low_spender     9.0
61       82       6.441221   Low_spender     9.0
62       82       6.441221   Low_spender     9.0
63       82       6.441221   Low_spender     9.0
64       82       6.441221   Low_spender     9.0
65       82       6.441221   Low_spender     9.0
66       82       6.441221   Low_spender     9.0
67       98       8.028000   Low_spender     9.0
68       98       8.028000   Low_spender     9.0
69       98       8.028000   Low_spender     9.0
70       98       8.028000   Low_spender     9.0
71       98       8.028000   Low_spender     9.0
72       98       8.028000   Low_spender     9.0
73       98       8.028000   Low_spender     9.0
74      109       5.850725   Low_spender     9.0
75      120       9.385714   Low_spender     9.0
76      120       9.385714   Low_spender     9.0
77      185       7.715000   Low_spender     9.0
78      195       7.149730   Low_spender     9.0
79      195       7.149730   Low_spender     9.0
80      195       7.149730   Low_sp

### question 7

In [40]:
# Creating a column for frequet users
ords_prods_merged_updated['Customer_frequency'] = ords_prods_merged_updated.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

C:\Users\amany\AppData\Local\Temp\ipykernel_1988\2157578290.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ords_prods_merged_updated['Customer_frequency'] = ords_prods_merged_updated.groupby(['user_id'])['days_since_prior_order'].transform(np.median)


In [41]:
# Creating flag on order frequency
ords_prods_merged_updated.loc[ords_prods_merged_updated['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'


C:\Users\amany\AppData\Local\Temp\ipykernel_1988\109097720.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ords_prods_merged_updated.loc[ords_prods_merged_updated['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'


In [42]:
ords_prods_merged_updated.loc[(ords_prods_merged_updated['Customer_frequency']>10) & (ords_prods_merged_updated['Customer_frequency']<=20),'frequency_flag'] = 'Regular customer'


In [43]:
ords_prods_merged_updated.loc[ords_prods_merged_updated['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [44]:
# Check frequency flags

ords_prods_merged_updated['frequency_flag'].value_counts(dropna = False)

Frequent Customer        21576934
Regular customer          7217005
Non_frequent_customer     3639575
Name: frequency_flag, dtype: int64

In [45]:
ords_prods_merged_updated[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(100)

user_id  days_since_prior_order  Customer_frequency         frequency_flag
0         1                     NaN                20.5  Non_frequent_customer
1         1                    15.0                20.5  Non_frequent_customer
2         1                    21.0                20.5  Non_frequent_customer
3         1                    29.0                20.5  Non_frequent_customer
4         1                    28.0                20.5  Non_frequent_customer
..      ...                     ...                 ...                    ...
95      360                     NaN                 4.0      Frequent Customer
96      377                     3.0                16.5       Regular customer
97      387                    22.0                 8.0      Frequent Customer
98      420                    30.0                 7.0      Frequent Customer
99      420                    16.0                 7.0      Frequent Customer

[100 rows x 4 columns]

## Exporting the Dataframe

In [52]:
ords_prods_merged_updated.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merged_updated.pkl'))